First EQC test for secondary indexing - 2.0 branch #822 [JIRA: RIAK-2185]

engelsanchez opened the pull request:
It needs more love to match the functionality of the regular 2i tests,
but this is a good beginning. It can show a race in 2i pagination that
has been in the code for years. The race does not show up in run in a
single node with very few vnodes. In my setup it shows up immediately
if configure to use 32 vnodes.  Ideally we could use
PULSE to make scheduling in the riak nodes less deterministic and help
find this kind of issue, but currently PULSE is a single node joint.

This test uses a different bucket for each EQC command run. It is not
perfect isolation, but it is good enough and very fast compared to
taking down the entire cluster in between each run.

engelsanchez committed:
First EQC test for secondary indexing

It needs more love to match the functionality of the regular 2i tests,
but this is a good beginning. It can show a race in 2i pagination that
has been in the code for years. The race does not show up in run in a
single node with very few vnodes. I tried 32.  Ideally we could use
PULSE to make scheduling in the riak nodes less deterministic and help
find this kind of issue, but currently PULSE is a single node joint.

This test uses a different bucket for each EQC command run. It is not
perfect isolation, but it is good enough and very fast compared to
taking down the entire cluster in between each run.

jonmeredith commented:
Thanks Engel,

EQC's are tricky beasts, and can be especially opaque when returning to them months and years later.  Would you mind adding a comment block at the top of the test explaining the structure of the state and how the test works.

It might be worth renaming the two indexing functions, from looking at it, I think 
index_objects1 adds same 2I entry to all objects
index_objects2 adds many 2I entries to single object.
but it would be nice to make that explicit.

After that we can merge. I've created a new issue to track the work needed to generalize the test, and would probably make a good project for EQC training (RIAK-2072).  Any comments on where to pick up would be good as TODOs in the code or added to the issue.

jonmeredith commented:
Pretty please.

engelsanchez commented:
Oh hey. I hadn't even seen this PR, but I'm currently working on documenting, refactoring and tweaking the test a bit more.

engelsanchez committed:
Clean up and improve

General clean up and organization, with an eye on making this a
template for other EQC tests.

Preparing code to handle more than the PB client.

engelsanchez commented:
And when I said "this PR", I meant this PR comment, not this PR I opened myself.
And when I said "I said", I mean I typed in a GH comment.

The last commit has a lot of refactoring for readability and documentation, as well as the beginnings of support for the HTTP client. Unfortunately the rest of the HTTP client support is becoming a quest to fix the client itself, since it is not properly encoding binary terms and the generator is quickly finding those. POS

engelsanchez commented:
I give up on the HTTP client. I can't even figure out how to properly update index tags on an object through rhc.  While debugging, I got into this case where the PB client shows me the index metadata for an object, but the http get does not. This happens even with brand new objects, consistently:

```
52> rp(riakc_pb_socket:get(C, <<"bucket265">>, <<"00011">>)).
{ok,{riakc_obj,<<"bucket265">>,<<"00011">>,
               <<107,206,97,96,96,96,204,96,202,5,82,28,71,131,54,114,75,
                 127,9,152,159,193,148,200,156,199,202,32,246,194,238,60,
                 95,22,0>>,
               [{{dict,4,16,16,8,80,48,
                       {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                       {{[],[],[],[],[],[],[],[],[],[],
                         [[<<"content-type">>,97,112,112,108,105,99,97,116,105,111,
                           110,47,111,99,116,101,116,45,115,116,114,101,97,109],
                          [<<"X-Riak-VTag">>,51,99,100,52,82,72,85,70,116,116,90,77,
                           107,103,54,89,54,82,65,121,69,109]],
                         [[<<"index">>,{<<"i1_int">>,<<"10">>}]],
                         [],
                         [[<<"X-Riak-Last-Modified">>|{1439,329302,728106}]],
                         [],[]}}},
                 <<"00011">>}],
               undefined,undefined}}
ok
53> rp(rhc:get(H, <<"bucket265">>, <<"00011">>)).
{ok,{riakc_obj,<<"bucket265">>,<<"00011">>,
               <<107,206,97,96,96,96,204,96,202,5,82,28,71,131,54,114,75,
                 127,9,152,159,193,148,200,156,199,202,32,246,194,238,60,
                 95,22,0>>,
               [{{dict,3,16,16,8,80,48,
                       {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                       {{[],[],[],[],[],[],[],[],[],[],
                         [[<<"content-type">>,97,112,112,108,105,99,97,116,105,111,
                           110,47,111,99,116,101,116,45,115,116,114,101,97,109],
                          [<<"X-Riak-VTag">>,34,51,99,100,52,82,72,85,70,116,116,90,
                           77,107,103,54,89,54,82,65,121,69,109,34]],
                         [],[],
                         [[<<"X-Riak-Last-Modified">>|{1439,329302,0}]],
                         [],[]}}},
                 <<"00011">>}],
               undefined,undefined}}
```

I'll check in the code with http disabled in the client type generator.

engelsanchez committed:
Fleshed out but disabled http client support

The http client seems to have a lot of problems and works differently
than the PB client.  Index values are not coming back on a get, even
though a PB get returns them, so indexing is not working properly.
Binaries are not encoded properly when the query index URL is created,
and empty binaries are taken to mean "no value".

engelsanchez commented:
@jonmeredith I'm done for now with changes to this PR. It seems to me the test is good as is for a first version, and the http client needs some love before we can use it in the test.  Can we have someone review this to merge? Let me know if there's anything more in terms of documentation that may be needed.

jonmeredith commented:
Holding off on merge until sext issue resolved.

engelsanchez commented:
Hey Jon. I don't see why. The fact that the test found a problem means that we should be using the test, not avoiding it.  Let's get this merged and in Giddyup, and the red will go away when that problem is fixed. We may find even more.

DSomogyi commented:
create jira issue

Basho-JIRA commented:
Created at Heather's request via email (Subject = Data Platform Riak Deltas, sent 10/9/2015 @ 3:23 PST). 

***[posted via JIRA by Derek Somogyi]***

Basho-JIRA commented:
[~jdaily], did this get included in the PR merge that happened with the other one? 

***[posted via JIRA by Derek Somogyi]***

Basho-JIRA commented:
[~jdaily], did this get included in the PR merge that happened with the other one (RTS-324)? 

***[posted via JIRA by Derek Somogyi]***

Basho-JIRA commented:
The PR to merge into mainline riak_test is still outstanding; separately I've just created https://github.com/basho/riak_test/pull/884 to merge into the `riak_ts` branch and YOLO merged it. 

***[posted via JIRA by John Daily]***

Basho-JIRA commented:
Note the test still fails because we haven't addressed the sext bug that pops up in very rare occasion (someone has to index very, very large negative integers, e.g. -5921785481) 

***[posted via JIRA by John Daily]***

macintux commented:
This should not have been closed. Reopening.

Commit:
Prevent large negative integers as values

This is due to a sext encoding bug present in current riak releases.
When sext and riak has been updated with a fix, revert this commit.

Commit:
Always ensure leveldb is the backend for 2i

JeetKunDoug commented:
:+1:
